[MRG+1] CookiesMiddleware: keep cookies from 'Cookie' request header, fix encoding

elacuesta opened the pull request:
Fixes #1992.

This PR modifies the `CookiesMiddleware._get_request_cookies` method to prevent the middleware from discarding cookies already set in the headers (either by the `DefaultHeadersMiddleware` or by using the `Request`'s `headers` constructor parameter).

Update: also fixes #3575

kmike commented on scrapy/downloadermiddlewares/cookies.py:
```diff
@@ -81,6 +82,14 @@ def _format_cookie(self, cookie):
     def _get_request_cookies(self, jar, request):
+        # from 'Cookie' request header
+        cookie_header = request.headers.get('Cookie', '')
+        cookie_list = re.split(';\s*', cookie_header.decode('latin1'))
```
Are there stdlib functions to parse cookies?

Also, cookies should be UTF-8 in most cases (but other encodings are also possible, so the code must be robust); .decode('latin1') never fails, but I wonder if non-ascii UTF-8 cookies are preserved correctly now. Could you please add a test for it?

kmike commented on tests/test_downloadermiddleware_cookies.py:
```diff
@@ -229,3 +229,17 @@ def test_local_domain(self):
         self.assertIn('Cookie', request.headers)
         self.assertEqual(b'currencyCookie=USD', request.headers['Cookie'])
+
+    def test_keep_cookie_header(self):
```
could you please also add a test which checks that it works with DEFAULT_HEADERS_MIDDLEWARE?

elacuesta replied:
Amended commit, test added 👍

elacuesta commented on scrapy/downloadermiddlewares/cookies.py:
```diff
@@ -71,7 +72,10 @@ def _debug_set_cookie(self, response, spider):
 
     def _format_cookie(self, cookie):
         # build cookie string
-        cookie_str = '%s=%s' % (cookie['name'], cookie['value'])
```
This is the part I'm least certain about. The tests were failing for py3 because the `b` char before a `bytes` object was being appended to the string, i.e.: `'%s=%s' % ('e', b'\xc3\xaa')`returned `"e=b'\\xc3\\xaa'"`, making the test comparison fail.
@kmike What do you think about it?

elacuesta replied:
It also works with `cookie_str = to_bytes('%s=%s' % (cookie['name'], to_native_str(cookie['value'])))` (see https://github.com/elacuesta/scrapy/commit/9a74d4257dea86fb72271edfd68ba750e9a2b995)

elacuesta commented:
@kmike, about
> cookies should be UTF-8 in most cases (but other encodings are also possible, so the code must be robust)

I believe the current `CookieJar` implementation (`scrapy.http.cookies.CookieJar`, based on `http.cookiejar`) does not handle encodings other than UTF8 properly. A different implementation could solve that, but IMHO that would fall out of the scope of this PR. 
Check the following snippet:

```
# -*- coding: utf-8 -*-

from __future__ import print_function
from scrapy.http import Request, Response
from scrapy.http.cookies import CookieJar

def test_cookie(encoding):
    print('# Encoding: {} #'.format(encoding))
    jar = CookieJar()
    response = Response('http://example.org', headers={'Set-Cookie': u'a=ä'.encode(encoding)})
    request = Request('http://example.org')
    cookies_from_response = jar.make_cookies(response, request)
    print('cookies:', cookies_from_response)
    for cookie in cookies_from_response:
        jar.set_cookie(cookie)
    jar.add_cookie_header(request)
    print('Cookie header:', request.headers['Cookie'])
    print()

test_cookie('utf8')
test_cookie('latin1')
test_cookie('utf16')
```
and the output:
```
$ python2 cookies_jar.py 
# Encoding: utf8 #
cookies: [Cookie(version=0, name='a', value='\xc3\xa4', port=None, port_specified=False, domain='example.org', domain_specified=False, domain_initial_dot=False, path='/', path_specified=False, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={}, rfc2109=False)]
Cookie header: a=ä

# Encoding: latin1 #
cookies: [Cookie(version=0, name='a', value='\xe4', port=None, port_specified=False, domain='example.org', domain_specified=False, domain_initial_dot=False, path='/', path_specified=False, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={}, rfc2109=False)]
Cookie header: a=

# Encoding: utf16 #
cookies: [Cookie(version=0, name='\xff\xfea\x00', value='\x00\xe4\x00', port=None, port_specified=False, domain='example.org', domain_specified=False, domain_initial_dot=False, path='/', path_specified=False, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={}, rfc2109=False)]
Cookie header: ��a=�

$ python3 cookies_jar.py 
# Encoding: utf8 #
cookies: [Cookie(version=0, name='a', value='ä', port=None, port_specified=False, domain='example.org', domain_specified=False, domain_initial_dot=False, path='/', path_specified=False, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={}, rfc2109=False)]
Cookie header: b'a=\xc3\xa4'

# Encoding: latin1 #
cookies: [Cookie(version=0, name='a', value='�', port=None, port_specified=False, domain='example.org', domain_specified=False, domain_initial_dot=False, path='/', path_specified=False, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={}, rfc2109=False)]
Cookie header: b'a=\xef\xbf\xbd'

# Encoding: utf16 #
cookies: [Cookie(version=0, name='��a\x00', value='\x00�\x00', port=None, port_specified=False, domain='example.org', domain_specified=False, domain_initial_dot=False, path='/', path_specified=False, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={}, rfc2109=False)]
Cookie header: b'\xef\xbf\xbd\xef\xbf\xbda\x00=\x00\xef\xbf\xbd\x00'
```

kmike commented:
@elacuesta yeah, it can't handle them because the encoding is unknown. But it shouldn't raise an exception even if a header is not utf-8 - there are byte sequences which can raise UnicodeDecodeError when decoded from utf-8; the whole response shouldn't fail in this case.

elacuesta commented:
Right, what do you think about the following then?
1) mention in the docs that cookies should be either unicode strings or UTF8-encoded bytes
2) catch `UnicodeDecodeError` when parsing cookies in `CookiesMiddleware` and log a warning
3) discard/keep (with probably bogus value) non-UTF8 cookies

kmike commented:
> mention in the docs that cookies should be either unicode strings or UTF8-encoded bytes

It won't work - it is website who defines what are cookies, we can't prescribe that. Non-utf8 cookies are likely bugs though.

> catch UnicodeDecodeError when parsing cookies in CookiesMiddleware and log a warning

This sounds good to me.

Warnings can be annoying if one scrapes a website which uses such cookies consistently, so it'd be nice to document how to disable these warnings, e.g. by pointing to https://doc.scrapy.org/en/latest/topics/logging.html#advanced-customization.

> discard/keep (with probably bogus value) non-UTF8 cookies

Currently non-UTF8 cookies are kept with bogus values - `errors='replace'` is used everywhere. Though not ideal, I haven't heard anyone complaining; probably it doesn't happen often, or maybe it happens in cases where non-lossy cookie values are not important.

elacuesta commented:
@kmike It seems to me that `http.cookiejar` works consistently in py2/py3 only with unicode strings, I added some code to decode cookies from the `Cookie` header (which is always `bytes` thanks to the `Headers` implementation) using `utf8` and falling back to `latin1` in case of `UnicodeDecodeError`.

With this implementation the `Cookie` header would always end up encoded with `utf8` even if the cookies were already passed as bytes using a different encoding. What do you think about that?

elacuesta commented:
I made some changes to the tests and to the docs to point to the logging configuration page.

@kmike could you please re-check this? The current status is:

Cookies set in the `Request.cookies` attribute
1) UTF8-encoded bytes: sent as they are, #3575 is also fixed
2) Non-UTF8-encoded bytes: sent as UTF8-encoded bytes - a warning is logged
3) Unicode string: encoded with UTF8

Cookies set in the `Cookie` request header (currently being ignored)
1) UTF8-encoded bytes: sent as they are
2) Non-UTF8-encoded bytes: sent as UTF8-encoded bytes - a warning is logged
3) Unicode string: encoded with UTF8

elacuesta commented:
Added a check and a test case to filter out invalid cookies (dicts with no `name` or `value` in the `Request.cookies` attribute)

dangra commented on scrapy/downloadermiddlewares/cookies.py:
```diff
@@ -69,26 +69,60 @@ def _debug_set_cookie(self, response, spider):
+        """
+        def get_cookies_from_header(jar, request):
+            cookie_header = request.headers.get('Cookie') or b''
+            cookie_list_bytes = re.split(b';\s*', cookie_header)
```
`_get_request_cookies` is on every request path, considering that we should precompile regexes and exit earlier if `Cookie` header is not set.

elacuesta replied:
Makes a lot of sense, thanks.

elacuesta replied:
@dangra Done, thanks for the suggestion!

Gallaecio commented on docs/topics/downloader-middleware.rst:
```diff
@@ -200,6 +200,11 @@ CookiesMiddleware
    send them back on subsequent requests (from that spider), just like web
    browsers do.
 
+   .. caution:: When non-UTF8 encoded byte sequences are passed to a ``Request``,
```
:lipstick: ``` :class:`~scrapy.http.Request` ```

elacuesta replied:
👌

Gallaecio commented on scrapy/downloadermiddlewares/cookies.py:
```diff
@@ -17,6 +18,7 @@ class CookiesMiddleware(object):
     def __init__(self, debug=False):
         self.jars = defaultdict(CookieJar)
         self.debug = debug
+        self.re_cookie_header = re.compile(b';\s*')
```
```suggestion
        self.re_cookie_header = re.compile(br';\s*')
```

Otherwise, I think it will result in a warning due to `\s`.

I also wonder if it would make sense to add `\s*` before the semicolon as well.

elacuesta replied:
Changed the approach, not using `re` anymore, `bytes.split` and `bytes.strip` are more efficient in this case.

elacuesta commented:
Since we're only splitting by one char (";") and ignoring whitespaces, it seems like using plain `bytes.split` and `bytes.strip` is faster than using regex.

```python
In [1]: import re
   ...:
   ...: re_split = re.compile(br'\s*;\s*')
   ...:
   ...: def cookies(n=4):
   ...:     values = [b'%i=%i' % (i, i*10) for i in range(10**n)]
   ...:     return b' ; '.join(values)
   ...:
   ...: %timeit re_split.split(cookies())
   ...: %timeit [c.strip() for c in cookies().split(b';')]
7.58 ms ± 39.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.5 ms ± 42.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
